util/process_wrapper/process_wrapper.cc

```diff
@@ -143,24 +142,74 @@ int PW_MAIN(int argc, const CharType* argv[], const CharType* envp[]) {
       for (++i; i < argc; ++i) {
         arguments.push_back(argv[i]);
       }
-      // after we consume all arguments we append the files arguments
-      for (const System::StrType& file_arg : file_arguments) {
-        arguments.push_back(file_arg);
-      }
     } else {
       std::cerr << "process wrapper error: unknow argument \"" << ToUtf8(arg)
                 << "\"." << '\n';
       return -1;
     }
   }
 
+  // Expand tokens into the file arguments (since some paths might embed
+  // `${pwd}` that we need to resolve it from the subsitution mappings).
+  for (System::StrType& file_arg : file_arguments) {
+    ReplaceTokens(file_arg, subst_mappings);
+  }
+
+  // Process the arguments, making sure that we don't have paths to param files
+  // within another param file (that is an arg composed of `@` + a path), which
+  // rustc doesn't support.
+  // (We process them in reverse so that we can append the param files to the
+  // front of `file_arguments` in the right order).
+  for (auto arg_it = arguments.rbegin(); arg_it != arguments.rend(); arg_it++) {
+    System::StrType& arg = *arg_it;
+    ReplaceTokens(arg, subst_mappings);
+
+    // Check whether this is a regular argument or a path to a param file.
+    if (arg.size() > 0 && arg.substr(0, 1) == PW_SYS_STR("@")) {
+      System::StrType path = arg.substr(1);
+
+      // Read the actual params off the param file so that we can 1) extract
+      // potential other param files inside and 2) substitute mappings.
+      System::StrVecType params;
+      if (!ReadFileToArray(path, params)) {
+        std::cerr << "Cannot read param file " << ToUtf8(path) << "\n";
+        return -1;
+      }
+
+      for (auto par_it = params.rbegin(); par_it != params.rend(); par_it++) {
+        System::StrType& param = *par_it;
+        ReplaceTokens(param, subst_mappings);
+
+        // If we find a path to a param file, move it to the top-level file
```

Comment:
Do we need this behavior? Could we just fail? I'm imagining a bug in rules_rust that produces nested param files and we wouldn't notice because process wrapper would recover from that. I also worry that moving param file args around can change the order of flags in a surprising way so I try to avoid it in general.

Reply:
That was also my first intuition but it turns out that we have a test that specifically checks that extra `@param_file_path` arguments are supported, by adding such an arg to the `rustc_flags`:
https://github.com/bazelbuild/rules_rust/blob/b16c26ba5faf1c58ebe94582afd20567ce676e6d/test/unit/location_expansion/location_expansion_test.bzl#L31

I wasn't able to gather the rationale for supporting this in the PR #809 or the associated issue #801, but there's a comment from you on that review that might be worth re-discussing here 😉 https://github.com/bazelbuild/rules_rust/pull/809/files#r662057446

Reply:
I think this explains the scenario: https://github.com/bazelbuild/rules_rust/issues/801#issuecomment-869226649. There are ways to reimplement that use case to not use param files, but I think I'm changing my mind, we should support nested param files. But I think the correct support should keep the relative order of flags. What do you think about flattening the param files, as in, if we find a path to a param file, we read the content of that file and use that instead of the `@nested_file` flag?

Reply:
Ah that makes sense. I like the idea of flattening params files, I've update the PR to do that.